File: amzn/Trees and Graphs/SymmetricTree.js
/** RECURSIVE
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
return mirror(root,root);
};

function mirror(n1,n2) {
    if(n1 === null && n2 === null) return true; 
    if(n1 === null || n2 === null) return false;
    
    return n1.val === n2.val && 
        mirror(n1.right,n2.left) &&
        mirror(n1.left,n2.right)
}

File: amzn/Trees and Graphs/golang/SymmetricTree.go
/** RECURSIVE
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    if (root == nil) {
        return true
    }
    return symmetric(root.Right,root.Left)
}

func symmetric(n1 *TreeNode,n2 *TreeNode) bool {
    if(n1 == nil && n2 == nil) {
        return true
    }
    if(n1 == nil || n2 == nil) {
        return false
    }
    
    return n1.Val == n2.Val && symmetric(n1.Left,n2.Right) && symmetric(n1.Right,n2.Left)
}
